docs_app/content/guide/core-semantics.md

```diff
@@ -0,0 +1,25 @@
+# RxJS Core Semantics
+
+Starting in version 8, all RxJS operators that are provided in the core library MUST meet the following semantics. In the current version, version 7, all operators SHOULD meet the following semantics (as guidelines). If they do not, we need to track the issue on [GitHub](https://github.com/ReactiveX/rxjs/issues).
+
+## Purpose
+
+The purpose of these semantics is provide predictable behavior for the users of our library, and to ensure consistent behavior between our many different operators.
+
+## Operators
+
+    - MUST be a function that returns an [operator function](https://rxjs.dev/api/index/interface/OperatorFunction). That is `(source: Observable<In>) => Observable<Out>`.
+    - The observable returned by the operator function MUST subscribe to the source.
+    - If the operation performed by the operator can tell it not change anything about the output of the source, it MUST return the reference to the source. For example `take(Infinity)` or `skip(0)`.
+    - Operators that accept a "notifier", that is another observable source that is used to trigger some behavior, must accept any type that can be converted to an `Observable` with `from`. For example `takeUntil`.
+    - "Notifiers" that are provided to operators are expected to emit a value in order to trigger notification. Completion does not count as a notification.
+    - "Notifiers" provided directly to the operator MUST be subscribed to *before* the source is subscribed to. "Notifiers" created via factory function provided to the operator SHOULD be subscribed to at the earliest possible moment.
+    - The observable returned by the operator function is considered to be the "consumer" of the source. As such, the consumer MUST unsubscribe from the source as soon as it knows it no longer needs values before proceeding to do _any_ action.
+    - Events that happen after the completion of a source should happen after the source finalizes. This is to ensure that finalization always happens in a predictable time frame relative to the event.
+    - `Error` objects should never be retained longer than necessary. This is a possible source of memory pressure.
+    - `Promise` references should never be retained longer than necessary. This is a possible source of memory pressure.
+    - Operators that split a source `Observable<T>` into many child observables `Observable<Observable<T>>` should emit child observables that do not stop because the original consumer subscription is unsubscribed. This is because those child observables may be consumed outside of that subscription lifecycle. For example, a user could capture a grouped (child) observable emitted from `groupBy` and subscribe to it elsewhere. The purpose of the `groupBy` was to create observables, not dictate their lifespan.
```

Comment:
it would be helpful to explain when to ultimately unsubscribe from the source and how to "detect" this case.

Reply:
just working on a `window` style operator and while looking at the `window` source code I noticed that this operator family unsubscribed the nested observables immediately when the primary subscription is unsubscribed from:
https://stackblitz.com/edit/qsuarg?devtoolsheight=33&file=index.ts

Reply:
The more i dig into the window/groupBy use cases the more I think that it is much more intuitive if the group lifecycles are tied to the top-level subscription and consuming them outside of its context is an anti-pattern, if we talk about window/groupBy as operators. is there some widely used pattern that requires this?

i also tried to compare it to the refactorings done to the multicasting operators, where connect/share operators are also entirely handled within the primary subscription(s), while the connectable observable factory was a part of the multicast operator that was taken out of pipeline/operators into an observable factory so the `connect` method can be handled outside of the scope of a primary subscription.

Reply:
> how to "detect" this case.

If the operator takes a source of `Observable<A>` returns `Observable<Observable<B>>`, (windowX, groupBy) that's how you'd know. There's nothing to really "detect" in this case.

> The more i dig into the window/groupBy use cases the more I think that it is much more intuitive...

Yeah, I don't disagree. I'm putting this in here as a starting/discussion point. A lot of people don't even realize that this is a thing. So, I've added it to the core semantics document so the core team can agree on a behavior and move that way over the course of 7 -> 8 -> 9 versions. It very well might be the correct decision to stop `groupBy` from behaving this way. Or maybe we need to add some sort of "eject" functionality... either way, it's a point of discussion, and I'm glad you're picking up on it.

Reply:
cc @cartant  @kwonoj @kolodny @niklas-wortmann ☝️

Reply:
> There's nothing to really "detect" in this case.

by "detect" I meant, how does the operator find out when to unsubscribe from the source. I know how groupBy does it and window* doesn't do it, but I would expect a general guideline here on what the correct "detection mechanism" is.

----
After giving this some more thought on my side, I have to justify my original opinion on this topic.
Lets first talk about window*:
```ts
const sub1 = of(1, 2, 3)
  .pipe(
    windowCount(1),
    take(1),
    mergeAll()
  ).subscribe()
  // I would expect this to emit `1` (it currently does not).

let w: Observable<number>;
const sub2 = of(1, 2, 3)
  .pipe(
    windowCount(1),
    take(1),
  )
  .subscribe((v) => void (w = v));
w.subscribe();
// I would expect this to be UB,
// it will never emit in a sync case and may or may not work in the async case

let w: Observable<number>;
const sub3 = of(1, 2, 3)
  .pipe(
    windowCount(1),
    take(1),
  )
  .subscribe((v) => void v.subscribe());
  // this could work and emit `1` (it currently does not),
  // but I would consider this bad practice and should be discouraged.
  // if it eases implementation then UB would be fine too IMHO.
```

why should the first example work? the reason is the semantics of the inner observable. it has a clear definition of its lifecycle: the first window starts before the first source emission, emits the following 1 next notification, and then completes. since windowCount only operates on the sequence of notifications (and not the contents of notifications) and/or external notifiers (on other variants of the window operator) it is "predictable" what the inner observables will emit and I can imagine uses cases for chaining it with `take` et al.

this is also still in line with what I originally said: the observable is consumed within its "primary" subscription (referring to `sub1` here), as the nested/inner subscriptions are "invisible" to the user. this is not the case for the second example.

in the groupBy case it's basically the same, but I just don't see any use case for `pipe(groupBy(selector), take(n))` without having the possibility to control group openings. this was proposed in https://github.com/ReactiveX/rxjs/pull/6425. currently this can be worked around with something like `pipe(startWith(x, y, z), groupBy(selector), map(skip(1)), take(n))`, but it does not feel very nice.

Reply:
Why not just multicast the `groupBy()` Observable and manually connect it?

Reply:
@trxcllnt LMAO. I swear I didn't write `groupBy` to behave like this, I thought you did. I agree it's weird. I only recorded this semantic because I wanted to draw attention to it. I'd be much happier if it did not have that behavior.